### PR TITLE
Allow supplying the repo name to buildProject

### PR DESCRIPTION
**Pull request by h-lame**
In some cases the repository name for a project and the name of the
jenkins job that runs the build are different.  For example the
Contacts Admin publishing app lives in a repo called contacts-admin, but
the jenkins job is called contacts (as are many other puppet references
to this app).  This means that the contacts-admin app can't use the
standard buildProject method in its Jenkinsfile because this assumes that
the repo matches the job.

It's not 100% clear when we should use the repoName and when we should
use the jobName.  For example the publishGem and pushDockerImage methods
take a name, but it's not actually important that it's the repository or
jenkins job name because in neither case does it make a reference to
github or jenkins.  Where it's not obvious I've left the param as the
jobName to preserve existing behaviour.

### CODE REVIEW

**Comment by tijmenb** [comment]
This is probably a crazy suggestion, but how difficult would it be to solve the inconsistency and rename `contact` to `contacts-admin` everywhere?

**Comment by surminus** [comment]
I thought this was handled in the code that creates the Jenkins job configuration itself:

https://github.com/alphagov/govuk-puppet/blob/master/hieradata/common.yaml#L22-L23
https://github.com/alphagov/govuk-puppet/blob/master/modules/govuk_ci/manifests/job.pp

**Comment by h-lame** [comment]
@surminus - the job setup itself is fine and copes with repos having different names to the job.  This change means the Jenkinsfile run by the job can use the `buildProject` script instead of defining the whole job itself.  You can compare https://ci.integration.publishing.service.gov.uk/job/contacts/job/use-standard-jenkins-buildProject/1/ (which uses the existing version of `buildProject` which expects the repo-name and the job-name to match) and https://ci.integration.publishing.service.gov.uk/job/contacts/job/use-standard-jenkins-buildProject/2/ (which uses this version of `buildProject` and passes in the repo-name).  Yes both failed, but the first failed because it's looking for the wrong repo, the second failed because the app doesn't quite work under `buildProject` yet (I'm working on that though).

@tijmenb - it was my first plan, but I saw that lots of puppet-y things have `contacts` instead of `contact-admin` and I got scared.

FWIW: My ultimate goal is to have all the jobs in this list: https://github.com/alphagov/govuk-content-schemas/blob/master/Jenkinsfile#L32-L37 not be in that list and so all child jobs report themselves on github when we build govuk-content-schemas.  I expect repo-name vs job-name is gonna fix a handful of them, but not all.

**Comment by h-lame** [comment]
Double FWIW: https://ci.integration.publishing.service.gov.uk/job/contacts/job/use-standard-jenkins-buildProject/6/ is a successful build of contacts-admin using the version of buildProject from this PR.  I achieved this by letting the build fail then using the replay feature of jenkins to re-run that failed build and change the groovy scripts used.

File: modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy [hunk 1]
```diff
@@ -62,15 +62,22 @@
...
- *        - newStyleDockerTags. Tag docker images with timestamp and git SHA
+ *        - newStyleDockerTags Tag docker images with timestamp and git SHA
  *          rather than the default of the build number
+ *        - repoName Provide this if the Github Repo name for the app is
+ *          different to the jenkins job name.
  */
 def buildProject(Map options = [:]) {
 
-  repoName = JOB_NAME.split('/')[0]
+  jobName = JOB_NAME.split('/')[0]
```
[the remark this answers is not in the document]
Comment (h-lame) [reply modules/govuk_jenkins/files/var/lib/jenkins/groovy_scripts/govuk_jenkinslib.groovy]: That sounds very reasonable, I'll do a rebase.  Thanks!

**Pull request merged**